src/v2/auth.rs

```diff
@@ -35,91 +173,105 @@ impl Client {
+                }
 
-    /// Perform registry authentication and return an authenticated token.
-    ///
-    /// On success, the returned token will be valid for all requested scopes.
-    pub async fn login(&self, scopes: &[&str]) -> Result<TokenAuth> {
-        let subclient = self.clone();
-        let creds = self.credentials.clone();
-        let scope = scopes
-            .iter()
-            .fold("".to_string(), |acc, &s| acc + "&scope=" + s);
-
-        let token_ep = self.get_token_provider().await?;
-        let auth_ep = token_ep + scope.as_str();
-        trace!("login: token endpoint: {}", auth_ep);
-
-        let url = reqwest::Url::parse(&auth_ep).map_err(|e| {
-            Error::from(format!(
-                "failed to parse url from string '{}': {}",
-                auth_ep, e
-            ))
-        })?;
-
-        let auth_req = {
-            let auth_req = subclient.build_reqwest(Method::GET, url);
-            if let Some(creds) = creds {
-                auth_req.basic_auth(creds.0, Some(creds.1))
-            } else {
-                auth_req
-            }
-        };
+                let bearer_auth = r.json::<BearerAuth>().await?;
 
-        let r = auth_req.send().await?;
-        let status = r.status();
-        trace!("login: got status {}", status);
-        match status {
-            StatusCode::OK => {}
-            _ => return Err(format!("login: wrong HTTP status '{}'", status).into()),
-        }
+                match bearer_auth.token.as_str() {
+                    "unauthenticated" => bail!("token is unauthenticated"),
+                    "" => bail!("received an empty token"),
+                    _ => {}
+                };
 
-        let token_auth = r.json::<TokenAuth>().await?;
-        let mut t = token_auth.token().to_string();
+                // mask the token before logging it
+                let chars_count = bearer_auth.token.chars().count();
+                let mask_start = std::cmp::min(1, chars_count - 1);
+                let mask_end = std::cmp::max(chars_count - 1, 1);
+                let mut masked_token = bearer_auth.token.clone();
+                masked_token
+                    .replace_range(mask_start..mask_end, &"*".repeat(mask_end - mask_start));
 
-        if t == "unauthenticated" {
-            bail!("received token with value '{}'", t)
-        } else if t.is_empty() {
-            bail!("received an empty token")
+                trace!("authenticate: got token: {:?}", masked_token);
+
+                self.auth = Some(Auth::Bearer(bearer_auth));
+            }
         };
 
-        // mask the token before logging it
-        let chars_count = t.chars().count();
-        let mask_start = std::cmp::min(1, chars_count - 1);
-        let mask_end = std::cmp::max(chars_count - 1, 1);
-        t.replace_range(mask_start..mask_end, &"*".repeat(mask_end - mask_start));
+        if !self.is_auth().await? {
+            self.auth = None;
```

Comment:
I think this does not have any effect, but see comments above regarding doing a single assignment if possible.

Reply:
Ops, I removed this conditional altogether now as discussed in the stroked text at https://github.com/camallo/dkregistry-rs/pull/159#discussion_r445542963:

> It _[the `self.is_auth()` check]_ is a relic from the earlier code which double-checks that the authentication was successful. I'm confident enough to remove it. (removed because I thought this was on a different line)

 WDYT?

Reply:
Yes, I don't remember adding this for any specific reason, I think it was mostly an overly abundant amount of safety check.